### ISSUE TITLE
[DASH] Possibility of playback freezes when switching streams in the middle of `StreamingEngine` updates (`fetchAndAppend_()`).

### ISSUE THREAD

**Issue by JulianDomingo** [shaka-project/shaka-player#7156]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes.

**If the problem is related to FairPlay, have you read the tutorial?**

N/A.

**What version of Shaka Player are you using?**

`main`

**Can you reproduce the issue with our latest release version?**
Yes.

**Can you reproduce the issue with the latest code from `main`?**
Yes.

**Are you using the demo app or your own custom app?**
Custom cast receiver app for Paramount+.

**If custom app, can you reproduce the issue using our demo app?**
No, as constructing test content to reproduce this scenario is difficult outside of the test environment provided directly from Paramount+. 

It is a 3 hour DVR window, live, multiperiod, DRM, DAI, DASH stream.

**What browser and OS are you using?**
Chrome / CastOS (reproducible on all Cast devices).

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**
Google Nest Hub Max / Google Chromecast with TV (4K), latest [production firmware](https://support.google.com/chromecast/answer/7124014?hl=en).

**What are the manifest and license server URIs?**

N/A (this was reproduced using internal test content generated by Paramount+).

**What configuration are you using?  What is the output of `player.getConfiguration()`?**

```
{
  "abr": {
    "defaultBandwidthEstimate": 2000000,
  },
  "drm": {
    "retryParameters": {
      "maxAttempts": 4,
      "baseDelay": 400,
      "backoffFactor": 2,
      "timeout": 30000
    },
    "servers": {
      "com.widevine.alpha": [REDACTED]
    },
    "advanced": {
      "com.widevine.alpha": {
        "audioRobustness": "HW_SECURE_CRYPTO",
        "videoRobustness": "HW_SECURE_ALL"
      }
    }
  },
  "manifest": {
    "retryParameters": {
      "maxAttempts": 4,
      "baseDelay": 400,
      "backoffFactor": 2,
      "timeout": 30000
    },
    "dash": {
      "disableXlinkProcessing": true
    }
  },
  "streaming": {
    "rebufferingGoal": 10,
    "retryParameters": {
      "maxAttempts": 4,
      "baseDelay": 400,
      "backoffFactor": 2,
      "timeout": 30000
    },
    "stallThreshold": 5,
    "stallSkip": 1.1
  },
  "mediaSource": {
    "codecSwitchingStrategy": "reload"
  }
}
```

**What did you do?**

1. Cast the Paramount+ live content on any Cast device.

**What did you expect to happen?**
1. The live stream to play without playback freezing / stalling indefinitely.

**What actually happened?**

Playback freezes when an ABR switch occurs in the middle of an ongoing `StreamingEngine.fetchAndAppend_()`. 

To be more specific, when the above happens, the `switchInternal_()` logic (which gets triggered from the ABR switch) closes the `SegmentIndex` of the old stream:
https://github.com/shaka-project/shaka-player/blob/01545f4c3eb540d6cc32c11a91874d53ddf92386/lib/media/streaming_engine.js#L575-L581

In the context of DASH, eventually this leads to the clearing of the `references` array, which holds things like the `uris` of each segment:

https://github.com/shaka-project/shaka-player/blob/01545f4c3eb540d6cc32c11a91874d53ddf92386/lib/media/segment_index.js#L89

This can be problematic if the old stream's segment index closes before the `fetch()` happens, since everything in `reference` no longer exists:
https://github.com/shaka-project/shaka-player/blob/01545f4c3eb540d6cc32c11a91874d53ddf92386/lib/media/streaming_engine.js#L1672-L1673

As a result, a `PendingRequest` object with no URI will be created downstream:
https://github.com/shaka-project/shaka-player/blob/01545f4c3eb540d6cc32c11a91874d53ddf92386/lib/media/streaming_engine.js#L2587-L2597

https://github.com/shaka-project/shaka-player/blob/01545f4c3eb540d6cc32c11a91874d53ddf92386/lib/net/networking_engine.js#L326-L327

When the presentation time eventually reaches the media start time of the failed segment request, playback stalls.

---

Here is a snippet of the console logs illustrating the error I see:
![freeze_logs_explanation](https://github.com/user-attachments/assets/b3d61f68-a4ca-461c-9ddc-2903c0fda81d)

**Are you planning send a PR to fix it?**
Yes. The plan is to defer the closing of a stream's `SegmentIndex` to the next `onUpdate_()` call when an ABR switch happens during a update, similar to how Shaka can defer clearing of the buffer:

https://github.com/shaka-project/shaka-player/blob/01545f4c3eb540d6cc32c11a91874d53ddf92386/lib/media/streaming_engine.js#L595-L603

I will guard this change with a new config defaulting to turned off, since it's a core `StreamingEngine` change and may not necessarily be a problem on some platforms.

**Comment by joeyparrish** [comment]
A regression was reported in video-dev.org Slack, and they narrowed it down to the PR that closed this issue.  Playing https://d24rwxnt7vw9qb.cloudfront.net/v1/dash/e6d234965645b411ad572802b6c9d5a10799c9c1/All_Reference_Streams/4577dca5f8a44756875ab5cc913cd1f1/index.mpd with this PR, reportedly:

> playback stops rendering with a warning "cannot find segment endTime" just a few seconds prior to the rendering alt.

This was initially observed on Samsung TVs, but then later on LG and even in Chrome.

I'm reverting the PR and reopening the issue.

**Comment by joeyparrish** [comment]
Another note from the reporter:

> It happens on any content on our end as long as it had many variants. When the player starts bumping to the upper variant in resolution, after the buffer time is exhausted, the playback halts at this frame.

**Comment by JulianDomingo** [comment]
Thanks for reverting it Joey - I was able to reproduce it as well, and the root cause is due to only calling `segmentIndex.release()` within `handleDeferredCloseSegmentIndexes_()`, but not setting `segmentIndex = null;`.

Since checking the nullability of `segmentIndex` is how `StreamingEngine` determines if it should create a new `segmentIndex` for the active stream, the "cannot find segment endTime" error can commonly occur if:

```
1. We switch from stream A -> B during an update
2. Update finishes, so call A.release() only
3. Regular playback resumes...
4. We decide to switch back to A (regardless of whether an update is currently happening or not)
5. StreamingEngine sees A.segmentIndex is NOT null, since we never actually set it to null
6. StreamingEngine therefore never populates A.segmentIndex with new references, 
   so A.segmentIndex is completely empty for the rest of the playback session
7. We get the "cannot find segment endTime" error perpetually
```

---

IIRC the reason we decided to only call `release()` was from this [past conversation within the reverted PR](https://github.com/shaka-project/shaka-player/pull/7157/files/db34775b67c88e53945d118e2904cf2648b7d8e3#r1718735796). But my response was incorrect; even if we switch in an `A -> B -> A` manner during an update + considering the second paragraph statement, there will always only be one, distinct instance of `segmentIndex` for A. 

In other words, the initial iteration of the reverted PR would be the corrected version. I will send out another PR in a bit with better unit tests to check against this behavior.
